### PR TITLE
[CUDA] Set GPU device ID in threads

### PR DESCRIPTION
**Pull request by shiyu1994**
This is to fix the issue proposed in #6018. As reported in the issue, currently illegal memory access may arise when using a `gpu_device_id > 0` for `cuda` tree learner. This is because some CUDA memory is allocated in threads. Though the device ID is set in the main thread, it is not set in the newly created threads. The allocated CUDA memory in those threads may reside in GPU 0 by default, which is different with the `gpu_device_id`. Accessing to such memory on `gpu_device_id` may cause illegal memory access.

Here's one example,
https://github.com/microsoft/LightGBM/blob/20975badd7ea3083c89f08a1f4dfda1a9789f6bc/src/io/cuda/cuda_column_data.cpp#L120-L137

### CODE REVIEW

File: src/io/cuda/cuda_column_data.cpp [hunk 1]
```diff
@@ -224,6 +223,7 @@ void CUDAColumnData::ResizeWhenCopySubrow(const data_size_t num_used_indices) {
   #pragma omp parallel for schedule(static) num_threads(num_threads_)
   for (int column_index = 0; column_index < num_columns_; ++column_index) {
     OMP_LOOP_EX_BEGIN();
+    SetCUDADevice(gpu_device_id_, __FILE__, __LINE__);
```
Comment (guolinke) [inline src/io/cuda/cuda_column_data.cpp:226]: do we need to set this inside loop?

Comment (shiyu1994) [reply src/io/cuda/cuda_column_data.cpp:231]: Move outside. [d4695ff](https://github.com/microsoft/LightGBM/pull/6028/commits/d4695ff37be9cbd2b85de1617252e8e9e883e699)

**Review by guolinke — APPROVED** [review]

**Review by jameslamb — APPROVED** [review]
Thanks for the thorough explanation! It's really helpful for my understanding 😊

**Comment by jameslamb** [comment]
I just merged latest `master` into this, to get the fixes from #6032.

@shiyu1994 , could you go into the repo settings and check this "Always suggest updating pull request branches" box?

image

That would add a button that you can click on PRs to merge `master` in them, so there wouldn't be a need to pull and push locally. I've found that saves a bit of work in other projects I work on.

**Comment by shiyu1994** [comment]
> could you go into the repo settings and check this "Always suggest updating pull request branches" box?

Done.

**Pull request merged**